### DISCUSSION TITLE
Modelling DC-DC or DC Coupled Systems and Power Limits

### DISCUSSION THREAD

**Discussion by leejt489** [pvlib/pvlib-python#1837] [General]
Hi,

I'm exploring using this library in commercial software. We deal with a lot of microgrids and off-grid hybrid power systems with solar.

There are two modelling aspects prevalent in these systems that are not clear to me how to address with pvlib:

1. DC coupled systems - the PVSystem accepts an inverter, but in some cases, the array may be interfaced with a DC-DC converter rather than an AC-DC converter, especially when the PV system is connected to a battery.
2. Dynamic limits at the converter - whether the converter is DC-DC or DC-AC, most devices I am familiar with accept a power or current limit that can be dynamically set as part of a real-time controller. This has many applications from controlling the charge power to a local energy storage device, shaping the net load on behind-the-meter systems supervisory control of a PV plant (transmission or distribution).

I'm thinking that the simplest approach to 1) is to model the DC-DC converter as an inverter but treat the output as "DC" at a different voltage in my application and to 2) is to take the moment-by-moment min of the power output with the dynamic limit. Seems workable enough, but I imagine this has come up before, and am curious what others have implemented or thought about? I think some built-in support for DC-coupled systems and generalizing "inverter" to "converter" would make the package more applicable out-of-the-box to off-grid and hybrid systems and microgrids especially.

**Comment by mikofski** [comment]
There is some very simple dc-dc power conversion for dc-optimizers here: https://github.com/pvlib/pvlib-python/issues/1588

there’s also a battery model here:
- https://github.com/pvlib/pvlib-python/issues/1333
- https://github.com/pvlib/pvlib-python/pull/1378

**Comment by leejt489** [comment]
Thank you @mikofski. I think my main question is what would it take so that when creating a simple model chain, like in the intro tutorial, you could pass in a generic converter instead of the argument that is explicitly named `inverter`? I'm new to pvlib, but my sense is that whether the converter is DC-AC or DC-DC is not really substantive, it seems the main features of the model are around finding the MPP and applying efficiencies and that the waveform of the output doesn't really impact anything. It's just the the pvlib syntax explicitly refers to "ac" and "dc" outputs and an "inverter".

It seems like my best option for basic modelling is just to ignore the naming and pretend that my DC-DC converter is an inverter. Would you agree?

But for general use it seems like the library could benefit from abstracting the inverter to a more generic converter. It wasn't very hard to come up with a workaround, but it would just be nice to be able to use the out-of-the-box modelling as in the intro tutorial for a DC-coupled system. 

I feel like I need to understand a little better what that would involve before submitting a feature request.

**Reply by cwhanse** [reply]
> main features of the model are around finding the MPP and applying efficiencies 

Yes, but I'd describe the current inverter models as estimating a DC-AC power conversion efficiency _assuming_ operation at MPP. An equivalent DC-DC converter model would represent a loss (conversion efficiency) as a function of input DC voltage and current, and desired output DC voltage.

**Reply by mikofski** [reply]
I agree with Cliff:

> the current inverter models as estimating a DC-AC power conversion efficiency _assuming_ operation at MPP

That’s why [the gist I linked to](https://gist.github.com/mikofski/70f29b7ea8d7fa91b5e6e02551bf42d2) demonstrates combing IV curves and then converts them to a fixed DC voltage. There is a draft PR https://github.com/pvlib/pvlib-python/pull/1781 that proposes adding mismatch to pvlib that might be helpful. Then I agree with you some general framework for dc-dc converters would be nice
